lib/membrane/video_compositor/video_compositor.ex

```diff
@@ -0,0 +1,431 @@
+      |> Enum.find(fn %OutputState{pad_ref: Pad.ref(:output, id)} -> pad_id == id end)
+
+    output_stream_format = %Membrane.H264{
+      framerate: {state.framerate, 1},
+      alignment: :nalu,
+      stream_structure: :annexb,
+      width: output_width,
+      height: output_height
+    }
+
+    spec =
+      get_child({:rtp_receiver, pad_id})
+      |> via_out(Pad.ref(:output, ssrc), options: [depayloader: RTP.H264.Depayloader])
+      |> child({:output_processor, pad_id}, %Membrane.VideoCompositor.OutputProcessor{
+        output_stream_format: output_stream_format
+      })
+      |> bin_output(Pad.ref(:output, pad_id))
+
+    {[spec: spec], state}
+  end
+
+  @impl true
+  def handle_child_notification(_msg, _child, _ctx, state) do
+    {[], state}
+  end
+
+  @spec start_vc_server(port_number()) :: :ok
+  defp start_vc_server(vc_port) do
+    architecture = system_architecture() |> Atom.to_string()
+
+    vc_app_path =
+      File.cwd!()
+      |> Path.join("video_compositor_app/#{architecture}/video_compositor/video_compositor")
+
+    spawn(fn ->
+      Rambo.run(vc_app_path, [], env: %{"MEMBRANE_VIDEO_COMPOSITOR_API_PORT" => "#{vc_port}"})
+    end)
+
+    started? =
+      0..50
+      |> Enum.reduce_while(false, fn _i, _acc ->
+        sleep_time_ms = 100
+        :timer.sleep(sleep_time_ms)
+
+        case VcReq.send_custom_request(%{}, vc_port) do
+          {:ok, _} -> {:halt, true}
+          {:error, _} -> {:cont, false}
+        end
+      end)
+
+    unless started? do
+      raise "Failed to startup and connect to VideoCompositor server."
+    end
+
+    :ok
+  end
+
+  @spec system_architecture() :: :darwin_aarch64 | :darwin_x86_64 | :linux_x86_64
+  defp system_architecture() do
+    case :os.type() do
+      {:unix, :darwin} ->
+        system_architecture = :erlang.system_info(:system_architecture) |> to_string()
+
+        cond do
+          Regex.match?(~r/aarch64/, system_architecture) ->
+            :darwin_aarch64
+
+          Regex.match?(~r/x86_64/, system_architecture) ->
+            :darwin_x86_64
+
+          true ->
+            raise "Unsupported system architecture: #{system_architecture}"
+        end
+
+      {:unix, :linux} ->
+        :linux_x86_64
+
+      os_type ->
+        raise "Unsupported os type: #{os_type}"
+    end
+  end
+
+  @spec register_input_stream(input_id(), State.t(), port_number()) ::
+          port_number()
+  defp register_input_stream(input_id, state, input_port \\ 4000) do
+    if state |> State.used_ports() |> MapSet.member?(input_port) do
+      register_input_stream(input_id, state, input_port + 2)
+    else
+      case VcReq.register_input_stream(input_id, input_port, state.vc_port) do
+        :ok ->
+          input_port
+
+        {:error, %Req.Response{}} ->
+          register_input_stream(
+            input_id,
+            state,
+            input_port + 2
+          )
+
+        _other ->
+          raise "Register input failed"
+      end
+    end
+  end
```

Comment:
Will this function be called many times? If yes, you will check port `4000` many times. This could be fixed by adding some randomisation.

BTW, why the first port, that you try to open, is 4000? It doesn't sound like a good idea